desc/compute/utils.py

```diff
@@ -0,0 +1,272 @@
+        X expanded to match the grid's pattern.
+
+    """
+    assert surface_label in {"rho", "theta", "zeta"}
+    if surface_label == "rho":
+        assert len(x) == grid.num_rho
+        return x[grid.inverse_rho_idx]
+    if surface_label == "theta":
+        assert len(x) == grid.num_theta
+        return x[grid.inverse_theta_idx]
+    if surface_label == "zeta":
+        assert len(x) == grid.num_zeta
+        return x[grid.inverse_zeta_idx]
+
+
+def surface_integrals(grid, q=1, surface_label="rho", match_grid=False):
+    """Compute the surface integral of a quantity for all surfaces in the grid.
+
+    Parameters
+    ----------
+    grid : Grid
+        Collocation grid containing the nodes to evaluate at.
+    q : ndarray
+        Quantity to integrate.
+    surface_label : str
+        The surface label of rho, theta, or zeta to compute integration over.
+    match_grid : bool
+        Whether to expand the result to match the dimension of the grid.
+        If False (default), the result is a single value for each surface in the grid.
+        If True, the result has repeated values to match the number of grid nodes.
+
+    Returns
+    -------
+    integrals : ndarray
+        Surface integrals of q over each surface in grid.
+
+    """
+    nodes, unique_idx, ds = _get_grid_surface(grid, surface_label)
+    max_surface_val = 1 if surface_label == "rho" else 2 * jnp.pi
+    bins = jnp.append(nodes[unique_idx], max_surface_val)
+    integrals = jnp.histogram(nodes, bins=bins, weights=ds * q)[0]
+
+    return expand(grid, integrals, surface_label) if match_grid else integrals
+
+
+def surface_averages(
+    grid,
+    q,
+    sqrt_g=jnp.array([1]),
+    surface_label="rho",
+    match_grid=False,
+    denominator=None,
+):
+    """Compute the surface average of a quantity for all surfaces in the grid.
+
+    Notes
+    -----
+        Implements the flux-surface average formula given by equation 4.9.11 in
+        W.D. D'haeseleer et al. (1991) doi:10.1007/978-3-642-75595-8.
+
+    Parameters
+    ----------
+    grid : Grid
+        Collocation grid containing the nodes to evaluate at.
+    q : ndarray
+        Quantity to average.
+    sqrt_g : ndarray
+        Coordinate system Jacobian determinant; see data_index["sqrt(g)"].
+    surface_label : str
+        The surface label of rho, theta, or zeta to compute the average over.
+    match_grid : bool
+        Whether to expand the result to match the dimension of the grid.
+        If False (default), the result is a single value for each surface in the grid.
+        If True, the result has repeated values to match the number of grid nodes.
+    denominator : ndarray
+        Volume over the other two coordinates besides surface_label.
+        This can be supplied to avoid redundant computations.
+
+    Returns
+    -------
+    averages : ndarray
+        Surface averages of q over each surface in grid.
+
+    """
+    if denominator is None:
+        if sqrt_g.size == 1:
```

Comment:
Was there a disadvantage to: `if jnp.ndim(sqrt_g) == 0 or len(sqrt_g) == 1:`? This would allow both float types and lists to work if supplied. Otherwise fine.

Reply:
Since this utility function should only ever be called with compute data, it's safe to assume it will always be an array. But otherwise there was no disadvantage. What is best practice here @f0uriest @dpanici?